### ISSUE TITLE
Zero downtime migration should disable queue

### ISSUE THREAD

**Issue by aygee** [babenkoivan/scout-elasticsearch-driver#132]
If migration is performed with queue enabled, the new index will be created and the alias is switched to the new index but the index may not be complete due to the queue.

I would suggest to disable queue when migration is performed.
Alternatively, queue the switching so that it is performed after all migration is finished.

**Comment by aygee** [comment]
The latter is probably easier by queueing the last two function calls in the `ElasticMigrateCommand`: `deleteSourceIndex()` and `createAliasForTargetIndex()`

**Comment by babenkoivan** [comment]
Hi @aygee, during migration the old index is used for reading and the new for writing.
If you are using queues for indexing, all data goes to the new index, that will be available right after migration is finished. I don't see a problem here, can you give me an example?

**Comment by aygee** [comment]
Hi @babenkoivan what I noticed is that the writing to the new index is queued, however the deleting of the old index and switching the alias to the new index are not queued. You can replicate this by running the migration on a large data and will find the data are missing on the alias until the queue is finished.

Or by turning of the queue worker so that it won't process the queue:

- Create a new index with some data
- Turn on queue in the scout config
- Don't run the queue worker (so that they stay in the queue)
- Run the migration
- You will find that the alias has been switched to new index, but does not have any data

Also on a separate note, the migration can be improved by:

- Having delete alias and create alias in one request.
- Delete the old index after switching the alias, not before.

Hope that makes sense. If not, happy to explain more.

**Comment by aygee** [comment]
@babenkoivan PR here: #139

**Comment by babenkoivan** [comment]
Hi @aygee, thanks for the explanation and sorry for taking so long to response!

I see now your concern, you expect to see fully ready to use index after migration, but it's filled with data only after some time. 

But do you think turning off the queue until migration is over is a good idea? It will make unpredictable load on some environments and can take a long time to run. Maybe we can make it as an option for the command. What do you think?

**Comment by aygee** [comment]
@babenkoivan I think the option to turn it off is a good idea. In my case, I would need the data to be available right after migration. But you're right, it may not be suitable for everyone.

I've created a pull request to address the above with some other improvements on #139 . I will add the flag `--no-queue` or something like that. What do you think?

**Comment by babenkoivan** [comment]
@aygee sounds great!

But I want to have a closer look at PR and unfortunately I'll have time to do that only during this weekend.

**Comment by aygee** [comment]
@babenkoivan I've updated the PR to add the `--no-queue` option. Please let me know what you think. Thanks :)

**Comment by aygee** [comment]
Hi @babenkoivan, any reasons why my PR was closed without merged? I have made the changes as you requested and also added comments for readability. I think my PR made the zero downtime migration to be a truly zero downtime which is great when you have large dataset.

**Comment by babenkoivan** [comment]
Hi @aygee!
Sorry, if I offended you, it wasn't my intention.
I've added a comment in the PR after the first review and you didn't reply for couple of weeks, that's why I closed it.

**Comment by aygee** [comment]
@babenkoivan No worries. I actually pushed the changes as you requested on 4 Oct and replied to your PR comment pretty much the next day after your comment. Maybe you missed it. Any chance you can review again please. See the reply in here #139. Thanks a lot.

**Comment by babenkoivan** [comment]
Hi @aygee, sorry for the late reply, I think you missed [my comment](https://github.com/babenkoivan/scout-elasticsearch-driver/pull/139#discussion_r222320458). I can't find your answer on it.

**Comment by aygee** [comment]
That is so weird! It says 'Pending', not sure what that means. When I clicked 'Resolve conversation' just now, it's gone! Not sure why you couldn't see it. Why is this so hard??

Anyway, my comments were:

1. I have updated to call the parent::getOptions - see the updated commit on https://github.com/babenkoivan/scout-elasticsearch-driver/pull/139/commits/cd15dc15a2846a9062e187b85bddd29e3edc3d0f

2. I feel the code is necessary to get a true zero downtime migration by doing a remove and add in a single transaction as described in elastic search documentation. Please let me know how I can make it better?

Should I create another PR?

**Comment by netsnatch** [comment]
Hi @aygee.
I faced the same problems as you.
I believe that doing atomic index replacements is the right way. And the option to turn off the queue is also convenient.

@babenkoivan, I have a question about zero downtime migration algorithm.
Why do you switch the writing to the new index and do the reading from the old index? If the migration process takes even a few minutes, users will not be able to find the data that has just been saved.

**Comment by aygee** [comment]
@netsnatch I did a PR #139 a while ago which implemented a truly zero downtime based on elastic search documentation by doing an atomic index replacement. Unfortunately the PR was rejected although I put a lot of effort updating the PR and answering the questions that @babenkoivan had. I'm no longer using this package so I'm not really up to date where this project is at. Feel free to have a look at the PR and maybe try to fix / get it into the project.

**Comment by iget-esoares** [comment]
@aygee I'm trying to give some attention to the PRs during the covid period. I'll take a look on your PR as soon as I stand on desktop tomorrow.
